lib/Dialect/LLHD/Transforms/Deseq.cpp

```diff
@@ -0,0 +1,1722 @@
+/// dependent nodes dirty.
+void Deseq::updateValue(Value value, ValueTable table) {
+  auto &slot = valueLattice[value];
+  if (slot != table) {
+    LLVM_DEBUG({
+      llvm::dbgs() << "- Update ";
+      value.printAsOperand(llvm::dbgs(), OpPrintingFlags());
+      llvm::dbgs() << " = " << table << "\n";
+    });
+    slot = table;
+    for (auto *user : value.getUsers())
+      markDirty(user);
+  }
+}
+
+/// Update the successor lattice values of a block, and mark all dependent nodes
+/// dirty.
+void Deseq::updateSuccessors(Block *block, SuccessorValues values) {
+  auto &slot = successorLattice[block];
+  for (auto [index, successor] : llvm::enumerate(block->getSuccessors())) {
+    if (!slot.empty() && slot[index] == values[index])
+      continue;
+    markDirty(successor);
+  }
+  slot = values;
+}
+
+/// Update the condition lattice value of a block, and mark all dependent nodes
+/// as dirty.
+void Deseq::updateCondition(Block *block, DNF condition) {
+  auto &slot = blockConditionLattice[block];
+  if (slot != condition) {
+    LLVM_DEBUG({
+      llvm::dbgs() << "- Update ";
+      block->printAsOperand(llvm::dbgs());
+      llvm::dbgs() << " condition = " << condition << "\n";
+    });
+    slot = condition;
+    markDirty(block->getTerminator());
+  }
+}
+
+/// Perform a data flow analysis of the values observed by the process' wait op,
+/// propagated from the past into the present as destination operands of the
+/// wait op, and values yielded as process results. This analysis distills out
+/// drive conditions as canonical DNF expressions and determines under which
+/// conditions given values are produced as process results. This information is
+/// then used to detect if a process expresses a pos/neg edge trigger detection,
+/// among other things.
+void Deseq::propagate() {
+  // Seed the lattice for any value defined outside the process.
+  SmallPtrSet<Operation *, 8> opsAboveSeen;
+  mlir::visitUsedValuesDefinedAbove(
+      process->getRegions(), [&](auto *opOperand) {
+        auto value = opOperand->get();
+        if (auto *defOp = value.getDefiningOp();
+            defOp && !observedBooleans.contains(value)) {
+          if (opsAboveSeen.insert(defOp).second)
+            propagate(defOp);
+        } else {
+          if (value.getType().isSignlessInteger(1))
+            updateBoolean(value, DNF(value));
+          updateValue(value, ValueTable(value));
+        }
+      });
+
+  // Seed the lattice for all operations in the process body.
+  for (auto &block : process.getBody()) {
+    propagate(&block);
+    for (auto &op : block)
+      propagate(&op);
+  }
+
+  // Propagate lattice values.
+  while (!dirtyNodes.empty()) {
+    auto node = dirtyNodes.pop_back_val();
+    if (auto *op = dyn_cast<Operation *>(node))
+      propagate(op);
+    else
+      propagate(cast<Block *>(node));
+  }
+  LLVM_DEBUG(llvm::dbgs() << "- Finished propagation\n");
+}
```

Comment:
Computing the DNF for all values is quite inefficient as the runtime is exponential in the number of terms. I see two options:
* Change the DNF datastructure to just collect the terms and only reorder them to form a DNF and optimize it when asked to do so instead of on-the-fly as soon as a new term is added
* Construct the DNFs backward from values that we need to know the DNF of

But we can just merge this and do some benchmarking/profiling as well as collecting statistics about expression sizes and number of primitive values to properly optimize this.

Reply:
Excellent point. I initially had an implementation that performed a depth-first traversal of the interesting signals to determine DNFs for the trigger signals specifically. I think I'll change Deseq over to that approach again in a second iteration. The pass currently has pretty poor scaling for exactly the reasons you mention.